### PR TITLE
Clear indexes while collecting merge tasks

### PR DESCRIPTION
**Pull request by ahmetmircik**
fixes https://github.com/hazelcast/hazelcast/issues/13456

ee pr to add test : https://github.com/hazelcast/hazelcast-enterprise/pull/2272

### CODE REVIEW

**Comment by mdogan** [comment]
Can you also add the test in the issue report after a little cleanup?

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMergeRunnable.java [hunk 1]
```diff
@@ -52,6 +54,23 @@
...
+    protected void onMergeRunnableConstruction(Map<String, Collection<RecordStore>> mergingStoresByName) {
+        super.onMergeRunnableConstruction(mergingStoresByName);
+
+        // Only clear global indexes here. Partitioned indexes will be cleared inside partition threads
+        for (Collection<RecordStore> recordStores : mergingStoresByName.values()) {
+            for (RecordStore recordStore : recordStores) {
+                MapContainer mapContainer = recordStore.getMapContainer();
+                Indexes indexes = mapContainer.getIndexes();
+                if (indexes != null) {
+                    indexes.clearIndexes();
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapMergeRunnable.java]: done

**Comment by ahmetmircik** [comment]
@mdogan, added the test.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java [hunk 2]
```diff
@@ -131,7 +131,7 @@ public void destroy() {
 
     @Override
     public void destroyInternals() {
-        clearIndexes();
+        clearPartitionedIndexes();
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java]: Probably it worth adding a comment why we are clearing only partitioned indexes here and where the global ones are cleared.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java [hunk 3]
```diff
@@ -257,7 +256,31 @@ protected void clearMapStore() {
         mapDataStore.reset();
     }
 
-    protected void clearIndexes() {
+    public void clearIndexes() {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java]: Seems like the method may be private.

Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java]: ... or at least protected to match the style of other cleanup methods.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java]: done

**Pull request merged**